### ISSUE TITLE
Support multi-tenancy for core BPMN process operations

### ISSUE THREAD

**Issue by koevskinikola** [camunda/zeebe#13279]
### Description

This issue covers multi-tenancy support for the core operations of BPMN processes in Zeebe. The following should be covered, i.e. any command/event records produced by the following actions should contain a `tenantId`:
* Starting a process instance
* Starting a process instance with results
* Starting a CallActivitiy
* Evaluating a Decision through a Business Rule task.
* Child instance creation.

#### Expected behaviors
The following behavior should be supported:
* A process instance is always owned by a tenant. A process instance can’t be shared across tenants.
* All child instances are owned by the same tenant as the process instance. That means that any `ProcessInstanceRelated` records are owned by the same tenant as the process instance.
* A `CallActivity` is started from a process definition with the following criteria:
   1. The process definition is owned by the same tenant as the parent process instance.
   2. If nothing is found, an `Incident` is raised.
* A `Decision` is evaluated if it has the following criteria:
   1. The Decision is owned by the same tenant as the parent process instance.
   2. If nothing is found, an `Incident` is raised.
* Filtering process definitions by version remains the same as the non-multi-tenant approach, i.e. if no version is specified, the latest version of the process definition owned by the tenant is used. (provided by #13238)
* A process instance can be started only if a tenant id is provided with the client request. The following scenarios are possible:
   * A tenant id is provided:
      1. A process instance is started if the user making the request has access to the tenant owning the process definition.
      2. The request/command is rejected if the user doesn't have correct access to the tenant.
   * A tenant id isn't provided:
      1. A process instance is started if the user only has access to a single tenant, and a single process definition can be found that can be mapped to that tenant.
      2. The request/command is rejected if the user has access to multiple tenants or no process definition is found.

:information_source: In the implementation, first focus on implementing the simple path, with a `tenantId` explicitly provided in the client request/command record. The goal is to deliver an MVP as soon as possible.

```[tasklist]
### Task breakdown
- [x] Engine can process multi-tenant ProcessInstanceCreationRecord commands with tenant id
- [x] Engine propagates tenant id to process instance elements.
- [x] Engine can start CallActivity by tenant id
- [ ] https://github.com/camunda/zeebe/issues/14211
- [ ] https://github.com/camunda/zeebe/issues/13536
- [ ] Notify Operate and Optimize teams that core BPMN process events are available from Zeebe
- [ ] ~CreateProcessInstance* RPCs documentation is expanded with multi-tenancy~
- [ ] ~Engine can evaluate Business Rule task Decision by tenant id~ (moved to DMN)
- [ ] ~Multi-tenancy data migration is provided for `ElementInstanceState`, `BannedInstanceState`~ WON'T DO
```

### Blocked by
- https://github.com/camunda/zeebe/issues/13238

**Comment by koevskinikola** [comment]
The task `Engine can evaluate Business Rule task Decision by tenant id` was moved to the https://github.com/camunda/zeebe/issues/13282 issue, as it fits better in the DMN scope of this Epic.

**Comment by koevskinikola** [comment]
The task `CreateProcessInstance* RPCs documentation is expanded with multi-tenancy` has been transfered to https://github.com/camunda/camunda-platform-docs/issues/2561.

**Comment by koevskinikola** [comment]
The task `Engine can start CallActivity by tenant id` was already completed in https://github.com/camunda/zeebe/pull/14328.

**Comment by korthout** [comment]
While we don't migrate `ElementInstanceState` and `BannedInstanceState` to be `TenantAware` right now, we could always add this later. The benefit would be further isolation of data. While keys are unique, the data is accessible when the engine makes a mistake due to a bug. By further isolating the data, we'd avoid such cases. But this can be done in a future iteration.
